packages/server/src/plugin/subscriptionCallback/index.ts

```diff
@@ -0,0 +1,713 @@
+          // receive a heartbeat for it.
+          self.logger?.error(`\`complete\` request failed: ${e}`, id);
+        } finally {
+          self.requestsInFlight.delete(completeRequestPromise!);
+          // clean up the subscription (and heartbeat if necessary)
+          self.terminateSubscriptions([id], callbackUrl);
+        }
+      },
+    };
+
+    subscriptionObject.startConsumingSubscription();
+    const subscriptionInfo =
+      this.subscriptionInfoByCallbackUrl.get(callbackUrl);
+    if (!subscriptionInfo) {
+      this.logger?.error(
+        `No existing heartbeat found for ${callbackUrl}, skipping subscription`,
+      );
+    }
+    subscriptionInfo?.subscriptionsById.set(id, subscriptionObject);
+  }
+
+  // Sends the `complete` request to the router.
+  async completeRequest({
+    errors,
+    callbackUrl,
+    id,
+    verifier,
+  }: {
+    errors?: readonly GraphQLError[];
+    callbackUrl: string;
+    id: string;
+    verifier: string;
+  }) {
+    let completeRequest: Promise<Response> | undefined;
+    try {
+      const maybeWithErrors = errors?.length ? ` with errors` : '';
+      this.logger?.debug(
+        'Sending `complete` request to router' + maybeWithErrors,
+        id,
+      );
+      const completeRequest = fetch(callbackUrl, {
+        method: 'POST',
+        body: JSON.stringify({
+          kind: 'subscription',
+          action: 'complete',
+          id,
+          verifier,
+          ...(errors && { errors }),
+        }),
+        headers: { 'Content-Type': 'application/json' },
+      });
+      this.requestsInFlight.add(completeRequest);
+      await completeRequest;
+      this.logger?.debug('`complete` request successful', id);
+    } catch (e) {
+      this.logger?.error(`\`complete\` request failed: ${e}`, id);
+      // TODO: handle this error
+      throw e;
+    } finally {
+      this.requestsInFlight.delete(completeRequest!);
+    }
+  }
+
+  collectAllSubscriptions() {
+    return Array.from(this.subscriptionInfoByCallbackUrl.values()).reduce(
+      (subscriptions, { subscriptionsById }) => {
+        subscriptions.push(...Array.from(subscriptionsById.values()));
+        return subscriptions;
+      },
+      [] as SubscriptionObject[],
+    );
+  }
+
+  async cleanup() {
+    // Wait for our inflight heartbeats to finish - they might handle cancelling
+    // some subscriptions
+    await Promise.all(
+      Array.from(this.subscriptionInfoByCallbackUrl.values()).map(
+        async ({ heartbeat }) => {
+          clearInterval(heartbeat?.interval);
+          await heartbeat?.queue[heartbeat.queue.length - 1];
+        },
+      ),
+    );
+    // Cancel / complete any still-active subscriptions
+    await Promise.all(
+      this.collectAllSubscriptions()
+        .filter((s) => !s.cancelled)
+        .map((s) => s.completeSubscription()),
+    );
+    // Wait for any remaining requests to finish
+    await Promise.all(this.requestsInFlight.values());
```

Comment:
Consider using `Promise.allSettled` rather than `Promise.all`. The reason being is that if one of them were to resolve to failure (via max retries exceeded) then it would return immediately rather than waiting for the other promises. I would make the same change for other uses of `Promise.all` in `cleanup`

Reply:
[c0faf33](https://github.com/apollographql/apollo-server/pull/7617/commits/c0faf33bba01e75d4c8a77b5a5c4ed219322de42)